parse.go

```diff
@@ -0,0 +1,395 @@
+		case itemLiteral:
+			matcher.Variants = append(matcher.Variants, Variant{
+				Key: p.parseVariantKey(),
+				QuotedPattern: QuotedPattern{
+					Patterns: p.parsePatterns(),
+				},
+			})
+			// TODO: Error handling: unexpected token
+		}
+	}
+
+	return matcher
+}
+
+func (p *parser) parseVariantKey() VariantKey {
+	value := p.current().val
+	if value == "*" {
+		return WildcardKey{Wildcard: '*'}
+	}
+
+	var key VariantKey
+
+	// Try to parse the literal as a number, if it fails then it is a name literal.
+	switch num := strAsNumber(value).(type) {
+	case nil:
+		key = LiteralKey{Literal: UnquotedLiteral{Value: NameLiteral{Name: value}}}
+	case int64:
+		key = LiteralKey{Literal: UnquotedLiteral{Value: NumberLiteral[int64]{Number: num}}}
+	case float64:
+		key = LiteralKey{Literal: UnquotedLiteral{Value: NumberLiteral[float64]{Number: num}}}
+	}
+
+	return key
+}
+
+func (p *parser) parseOption() Option {
+	var identifier Identifier
+
+	for item := p.current(); item.typ != itemExpressionClose; item = p.next() {
+		switch item.typ {
+		case itemOption:
+			identifier = p.parseIdentifier()
+
+		case itemLiteral:
+			option := LiteralOption{Literal: p.parseLiteral(), Identifier: identifier}
+
+			return option
+		case itemVariable:
+			option := VariableOption{Variable: Variable(p.current().val[1:]), Identifier: identifier}
+
+			return option
+
+			// TODO: Error handling: unexpected token
+		}
+	}
+
+	// todo: error. Reason: value is missing for option
+	return nil
+}
+
+func (p *parser) parseLiteral() Literal {
+	// If there is prefix "$" then it is unquoted name literal.
+	if strings.HasPrefix(p.current().val, "$") {
+		return UnquotedLiteral{Value: NameLiteral{Name: p.current().val[1:]}}
+	}
+
+	var literal Literal
+
+	// Try to parse the literal as a number, if nil then it is a quoted literal, otherwise it is a number literal.
+	switch num := strAsNumber(p.current().val).(type) {
+	case nil:
+		literal = QuotedLiteral{Value: p.current().val}
+	case int64:
+		literal = UnquotedLiteral{Value: NumberLiteral[int64]{Number: num}}
+	case float64:
+		literal = UnquotedLiteral{Value: NumberLiteral[float64]{Number: num}}
+	}
+
+	return literal
+}
+
+func (p *parser) parseFunction() Function {
+	return Function{Prefix: rune(p.current().val[0]), Identifier: p.parseIdentifier()}
+}
+
+func (p *parser) parseIdentifier() Identifier {
+	full := strings.Split(p.current().val, ":")
+
+	var (
+		ns   string
+		name string
+	)
+
+	switch len(full) {
+	// no namespace
+	case 1:
+		name = full[0]
+	// namespace + name
+	case 2:
+		ns = full[0]
+		name = full[1]
+	// edge case for ":namespace:function"
+	case 3:
+		ns = full[1]
+		name = full[2]
+		// TODO: error handling: case unexpected length
+	}
+
+	return Identifier{Namespace: ns, Name: name}
+}
+
+// helpers
+
+// strAsNumber tries to parse a string as a number.
+// Returns int64 or float64 if successful, otherwise returns nil.
+func strAsNumber(s string) interface{} {
```

Comment:
Number literal has the following definition. 

```abnf
; number-literal matches JSON number
; https://www.rfc-editor.org/rfc/rfc8259#section-6
number-literal = ["-"] (0 / ([1-9] *DIGIT)) ["." 1*DIGIT] [%i"e" ["-" / "+"] 1*DIGIT]
```

`strconv.ParseInt` does not parse according to the given RFC (it can parse `inf` and other values that are not part of the RFC). Please use `json.Unmarshal()` instead.

Always unmarshal to `float64`